divi/src/test/FakeWallet.cpp

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2021 The Divi developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "test/FakeWallet.h"
+
+#include "chain.h"
+#include "merkletx.h"
+#include "primitives/transaction.h"
+#include "random.h"
+#include "script/script.h"
+#include "sync.h"
+#include "WalletTx.h"
+
+#include <sstream>
+#include <string>
+
+namespace
+{
+
+/** Returns a unique wallet filename.  */
+std::string GetWalletFilename()
+{
+  static int cnt = 0;
+  std::ostringstream res;
+  res << "currentWallet" << (++cnt) << ".dat";
+  return res.str();
+}
+
+/** Returns a HD seed for testing.  */
+const CHDChain& getHDWalletSeedForTesting()
+{
+  static CCriticalSection cs_testing;
+  static bool toBeConstructed = true;
+  static CHDChain hdchain;
+  LOCK(cs_testing);
+
+  if (toBeConstructed)
+  {
+    CWallet wallet("test_wallet.dat");
+    wallet.defaultKeyPoolTopUp = 1;
+    bool firstLoad = true;
+    wallet.LoadWallet(firstLoad);
+    wallet.GenerateNewHDChain();
+    wallet.GetHDChain(hdchain);
+    toBeConstructed = false;
+  }
+
+  return hdchain;
+}
+
+/** Creates a "normal" transaction paying to the given script.  */
+CMutableTransaction createDefaultTransaction(const CScript& defaultScript, unsigned& index,
+                                             const CAmount amount)
+{
+  static int nextLockTime = 0;
+  const int numberOfIndices = GetRand(10) + 1;
+  index = GetRand(numberOfIndices);
+
+  CMutableTransaction tx;
+  tx.nLockTime = nextLockTime++; // so all transactions get different hashes
+  tx.vout.resize(numberOfIndices);
+  for(CTxOut& output: tx.vout)
+  {
+    output.nValue = 1;
+    output.scriptPubKey = CScript() << OP_TRUE;
+  }
+  tx.vout[index].nValue = amount;
+  tx.vout[index].scriptPubKey = defaultScript;
+  tx.vin.resize(1);
+  // Avoid flagging as a coinbase tx
+  tx.vin[0].prevout = COutPoint(GetRandHash(),static_cast<uint32_t>(GetRand(100)) );
+
+  return tx;
+}
+
+} // anonymous namespace
+
+FakeWallet::FakeWallet()
+  : FakeBlockIndexWithHashes(1, 1600000000, version),
+    CWallet(GetWalletFilename()),
+    fakeChain(this)
+{
+  Init();
+}
+
+FakeWallet::FakeWallet(FakeBlockIndexWithHashes& c)
+  : FakeBlockIndexWithHashes(1, 1600000000, version),
+    CWallet(GetWalletFilename()),
+    fakeChain(&c)
+{
+  Init();
+}
```

Comment:
I agree it would be nicer in the code to not inherit, but unfortunately we have to.  The reason is that with the later chance where `CWallet`'s constructor accepts the chain and block map, we need to have access to them already when we call the `CWallet` superclass constructor.  And the only way to do that is to inherit also from `FakeBlockIndexWithHashes`, so we can construct that instance before `CWallet`.

(An alternative would be to not inherit from `CWallet` either and have it as an instance variable, but I think that would make the tests less nice.  Having an instance that is directly a "real wallet" is IMHO good from the user's point of view, which is more important here.)

Reply:
> I agree it would be nicer in the code to not inherit, but unfortunately we have to. The reason is that with the later chance where `CWallet`'s constructor accepts the chain and block map, we need to have access to them already when we call the `CWallet` superclass constructor. And the only way to do that is to inherit also from `FakeBlockIndexWithHashes`, so we can construct that instance before `CWallet`.

You defined two constructors, what I'm saying is to ditch the default constructor and use only the constructor where we would be supplying an instance of `FakeBlockIndexWithHashes`. The instance can be supplied by the test fixture, not the `FakeWallet` class directly (which as you point out would then need to forcefully inherit). The `FakeWallet` in that case would not be stand alone, but would avoid the indirection. Based on that you can then create a very shallow wrapper that instantiates a `FakeBlockIndexWithHashes` and `FakeWallet` and has an accessor for the `FakeWallet` as a `CWallet&`.

Reply:
I see what you mean.  I personally think that the private inheritance from the `FakeBlockIndexWithHashes` is a bit hackish but an implementation detail, and then allows the best and simplest possible use of the class in external code, which is what matters most.

But yeah, defining a `FakeBlockIndexWithHashes` in the tests themselves is also not too much of an extra effort, so I will change the PR accordingly.

Reply:
Updated.